# Предобработка данных

- Автор:Ососов Ян
- Дата:31.05.2025

### Цели и задачи проекта

Познакомиться с данными, проверить их корректность и провести предобработку, получив необходимый срез данных.

### Описание данных

- Данные ` ` содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:
- 


### Содержимое проекта

- Загрузка данныых и знакомситво с ними
- Проверка ошибок и предобработка
- Фильтрация данных
- Кактегоризация данных
- Итоговые выводы

## 1. Загрузка данных и знакомство с ними



импортируем библиотекку `pandas` и данные датасета

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('')

Знакомимся с данными

In [3]:
print(df.head())
df.info()
print(df.shape)

                       Name Platform  Year of Release         Genre  NA sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  EU sales JP sales  Other sales  Critic Score User Score Rating  
0    28.96     3.77         8.45          76.0          8      E  
1     3.58     6.81         0.77           NaN        NaN    NaN  
2    12.76     3.79         3.29          82.0        8.3      E  
3    10.93     3.28         2.95          80.0          8      E  
4     8.89    10.22         1.00           NaN        NaN    NaN  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #

Мы имеем дело с данными в виде таблицы, у которой 11 столбцов и 16956 строк.
Они соответствуют описанию.
- Пропуски

Хочется отметить что 1,4,5,6,7 столбцы не имеют пропусков что отлично
В 0-м столбце две пустые строки это важный столбец, игнорировать пропуски нельзя
В 3-м столбце тоже два пропуска, соответствуют ли они тем пропускам в 0-м столбце нужно проверить
Во 2-м столбце уже больше пропусков это год релиза, посмотри что будет на след этапах обработки
ну и остальные 8,9,10 столбцы много пропусков сильно много посмотрим в какой форме мы будем применять эти данные
- Типы данных

Почему то год имееет тип `float64` всему виной пропуски, некоторые продажи тип `object`что тоже не до конца корректно ведь тут нет пропусков, и оценки тут в целом ясно почему формат не тот что нужно

- Названия столбцов

ну видимо мы буем приводить к _snake_case_ что тут добавлять

---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма


In [4]:
print(df.columns)

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')


стиль меняем на snake_case!

In [5]:
df.columns = df.columns.str.lower()
df.columns = df.columns.str.replace(' ', '_')
print(df.columns)

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')


Двигаемся дальше

### 2.2. Типы данных


Отложим этап с типами, ибоо заведомо понятно что от части проблема тиипов проблема пропусков, чтобы не возвращатся к ним сделаем все вместе и позже а сейчас еще раз посмотрим на данные:

Ну тогда посмотрим почему eu_sales, jp_sales, user_score, я это делал понижее ну еще разок тогда

In [6]:
df['user_score'].unique()

array(['8', nan, '8.3', '8.5', '6.6', '8.4', '8.6', '7.7', '6.3', '7.4',
       '8.2', '9', '7.9', '8.1', '8.7', '7.1', '3.4', '5.3', '4.8', '3.2',
       '8.9', '6.4', '7.8', '7.5', '2.6', '7.2', '9.2', '7', '7.3', '4.3',
       '7.6', '5.7', '5', '9.1', '6.5', 'tbd', '8.8', '6.9', '9.4', '6.8',
       '6.1', '6.7', '5.4', '4', '4.9', '4.5', '9.3', '6.2', '4.2', '6',
       '3.7', '4.1', '5.8', '5.6', '5.5', '4.4', '4.6', '5.9', '3.9',
       '3.1', '2.9', '5.2', '3.3', '4.7', '5.1', '3.5', '2.5', '1.9', '3',
       '2.7', '2.2', '2', '9.5', '2.1', '3.6', '2.8', '1.8', '3.8', '0',
       '1.6', '9.6', '2.4', '1.7', '1.1', '0.3', '1.5', '0.7', '1.2',
       '2.3', '0.5', '1.3', '0.2', '0.6', '1.4', '0.9', '1', '9.7'],
      dtype=object)

Оп выловилди tbd to be determed только будт определенн ну в целом то же  что его нет, но с точки зрения гипотезы того что оценок меньшле у непопулярных игр - она имеет меесто быть. здесь же речь идет о том что это не просто nan надо поудамть что сделать с этим птом

In [7]:
df['eu_sales'].unique()

array(['28.96', '3.58', '12.76', '10.93', '8.89', '2.26', '9.14', '9.18',
       '6.94', '0.63', '10.95', '7.47', '6.18', '8.03', '4.89', '8.49',
       '9.09', '0.4', '3.75', '9.2', '4.46', '2.71', '3.44', '5.14',
       '5.49', '3.9', '5.35', '3.17', '5.09', '4.24', '5.04', '5.86',
       '3.68', '4.19', '5.73', '3.59', '4.51', '2.55', '4.02', '4.37',
       '6.31', '3.45', '2.81', '2.85', '3.49', '0.01', '3.35', '2.04',
       '3.07', '3.87', '3.0', '4.82', '3.64', '2.15', '3.69', '2.65',
       '2.56', '3.11', '3.14', '1.94', '1.95', '2.47', '2.28', '3.42',
       '3.63', '2.36', '1.71', '1.85', '2.79', '1.24', '6.12', '1.53',
       '3.47', '2.24', '5.01', '2.01', '1.72', '2.07', '6.42', '3.86',
       '0.45', '3.48', '1.89', '5.75', '2.17', '1.37', '2.35', '1.18',
       '2.11', '1.88', '2.83', '2.99', '2.89', '3.27', '2.22', '2.14',
       '1.45', '1.75', '1.04', '1.77', '3.02', '2.75', '2.16', '1.9',
       '2.59', '2.2', '4.3', '0.93', '2.53', '2.52', '1.79', '1.3', '2.6',
   

In [8]:
df['jp_sales'].unique()

array(['3.77', '6.81', '3.79', '3.28', '10.22', '4.22', '6.5', '2.93',
       '4.7', '0.28', '1.93', '4.13', '7.2', '3.6', '0.24', '2.53',
       '0.98', '0.41', '3.54', '4.16', '6.04', '4.18', '3.84', '0.06',
       '0.47', '5.38', '5.32', '5.65', '1.87', '0.13', '3.12', '0.36',
       '0.11', '4.35', '0.65', '0.07', '0.08', '0.49', '0.3', '2.66',
       '2.69', '0.48', '0.38', '5.33', '1.91', '3.96', '3.1', '1.1',
       '1.2', '0.14', '2.54', '2.14', '0.81', '2.12', '0.44', '3.15',
       '1.25', '0.04', '0.0', '2.47', '2.23', '1.69', '0.01', '3.0',
       '0.02', '4.39', '1.98', '0.1', '3.81', '0.05', '2.49', '1.58',
       '3.14', '2.73', '0.66', '0.22', '3.63', '1.45', '1.31', '2.43',
       '0.7', '0.35', '1.4', '0.6', '2.26', '1.42', '1.28', '1.39',
       '0.87', '0.17', '0.94', '0.19', '0.21', '1.6', '0.16', '1.03',
       '0.25', '2.06', '1.49', '1.29', '0.09', '2.87', '0.03', '0.78',
       '0.83', '2.33', '2.02', '1.36', '1.81', '1.97', '0.91', '0.99',
       '0.95', '2.0'

Есть unknown ну тоже ничего не дает обратим в nan

Ну с годом понятно почему та не инт а флоат - есть пропуски

### 2.3. Пропуски

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16956 non-null  object 
 6   jp_sales         16956 non-null  object 
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       10152 non-null  object 
 10  rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


Как писал раньше первый столбец оч плохо иметь пропуски глянем

In [10]:
print(df[df['name'].isna()])

      name platform  year_of_release genre  na_sales eu_sales jp_sales  \
661    NaN      GEN           1993.0   NaN      1.78     0.53      0.0   
14439  NaN      GEN           1993.0   NaN      0.00      0.0     0.03   

       other_sales  critic_score user_score rating  
661           0.08           NaN        NaN    NaN  
14439         0.00           NaN        NaN    NaN  


Как было предположено ранее два пропуска в жанре и в имени это две одинаковые строки, вообще безполезные строкки для анализа а с учетом того что год не попадает в нужный интервал для исследования от 2000 по 2013 смело удаляем чтобы не мешали

In [11]:
df = df[~df['name'].isna()]

смотрим что вышло

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16954 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16954 non-null  object 
 2   year_of_release  16679 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16954 non-null  float64
 5   eu_sales         16954 non-null  object 
 6   jp_sales         16954 non-null  object 
 7   other_sales      16954 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       10152 non-null  object 
 10  rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.6+ MB


Теперь год, тоже важный столбец ибо нам нужен срез по годам, а в результате ненужные данные могут повлиять на результатыы исследования. посмотрим чего внутри

In [13]:
print(df[df['year_of_release'].isna()])
print('процент пропусков')
print(round(df['year_of_release'].isna().sum()/ df.shape[0] * 100, 2))

                                    name platform  year_of_release  \
183                      Madden NFL 2004      PS2              NaN   
379                     FIFA Soccer 2004      PS2              NaN   
458           LEGO Batman: The Videogame      Wii              NaN   
477           wwe Smackdown vs. Raw 2006      PS2              NaN   
611                       Space Invaders     2600              NaN   
...                                  ...      ...              ...   
16609  PDC World Championship Darts 2008      PSP              NaN   
16641                      Freaky Flyers       GC              NaN   
16685                          Inversion       PC              NaN   
16695        Hakuouki: Shinsengumi Kitan      PS3              NaN   
16760                       Virtua Quest       GC              NaN   

              genre  na_sales eu_sales jp_sales  other_sales  critic_score  \
183          Sports      4.26     0.26     0.01         0.71          94.0   
379

Хочется сказать что процент пропусков меньше 2х процентов в целом быть может можно игнорироовать и не учитывать. В целом есть продажи - есть релиз либо релиз в нужный промежуток времени либо нет, те все игры сущетсвуют, однако нужно также понять сколько игр попадет во временные ворота и проводить сравнение с этим числом ибо есть игры в название которых есть дата релиза, пока считаю целесообразныым заполнить пропуски индикатором для того чтобы перейти к нужному формату, для хранения информации о годе целесообразно испльзовать формат int приведем к нему сейчас: заменим все на -1, приведем к инту с оптимальным размером

In [14]:
print(df.sort_values(by='na_sales').head(15))

                                                    name platform  \
16955                                Winning Post 8 2016      PSV   
12864        Hajime no Ippo Portable: Victorious Spirits      PSP   
12866  Katekyoo Hitman Reborn! DS: Flame Rumble Mukur...       DS   
10071  King of Colosseum (Red): Shin Nippon x Zen Nip...      PS2   
12872  Cid to Chocobo no Fushigi na Dungeon: Toki Was...       DS   
12873                                     Vitamin X to Z      PSP   
4538                                     DanceStar Party      PS3   
12877         Jacqueline Wilson's Tracy Beaker: The Game       DS   
4533                               NiGHTS into dreams...      SAT   
12878  Teenage Mutant Ninja Turtles II: Back from the...       GB   
12882   Real Bout Garou Densetsu Special: Dominated Mind       PS   
12883                                   Minna no Sukkiri      PSP   
12887                                    Chaos Rings III      PSV   
12889                  Kamen Rider

In [15]:
print(df.sort_values(by='na_sales').tail(15))

                         name platform  year_of_release         genre  \
32    Call of Duty: Black Ops     X360           2010.0       Shooter   
11              Mario Kart DS       DS           2005.0        Racing   
21           Super Mario Land       GB           1989.0      Platform   
4    Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
6       New Super Mario Bros.       DS           2006.0      Platform   
18          Super Mario World     SNES           1990.0      Platform   
7                    Wii Play      Wii           2006.0          Misc   
8   New Super Mario Bros. Wii      Wii           2009.0      Platform   
14         Kinect Adventures!     X360           2010.0          Misc   
3           Wii Sports Resort      Wii           2009.0        Sports   
2              Mario Kart Wii      Wii           2008.0        Racing   
5                      Tetris       GB           1989.0        Puzzle   
9                   Duck Hunt      NES           19

In [16]:
df['year_of_release']=df['year_of_release'].fillna(-1)
df['year_of_release'] = pd.to_numeric(df['year_of_release'], downcast='integer')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16954 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16954 non-null  object 
 2   year_of_release  16954 non-null  int16  
 3   genre            16954 non-null  object 
 4   na_sales         16954 non-null  float64
 5   eu_sales         16954 non-null  object 
 6   jp_sales         16954 non-null  object 
 7   other_sales      16954 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       10152 non-null  object 
 10  rating           10085 non-null  object 
dtypes: float64(3), int16(1), object(7)
memory usage: 1.5+ MB


На счет пропусков в оценках, их много относительно общего объема

In [17]:
print('critic_score')
print(round(df['critic_score'].isna().sum()/ df.shape[0] * 100, 2))
print('user_score')
print(round(df['user_score'].isna().sum()/ df.shape[0] * 100, 2))
print('rating')
print(round(df['rating'].isna().sum()/ df.shape[0] * 100, 2))

critic_score
51.39
user_score
40.12
rating
40.52


Очень большой процент что делать пока не понятно, но взять точну информацю взять неоткуда, поскольку все метрики имеют формат float то можно привести кк нему и без заполнения, хочется посмотреть на кол-во пропусков в разрезе нужной даты, быть может это проблема того что информация о старых играх или утеряна или на тот момент просто не оценивали как стали делать потом, скорее интересно посмотреть на данные еще раз после фильтрации особенно быть может на данные где одна из оценок утеряна в разрезе нужныых временных рамок

перед тем как перейти к преобразованию типов посмторим на предмет выбросов и тд

In [18]:
print(df[['na_sales', 'eu_sales', 'jp_sales', 'other_sales', 'critic_score', 'user_score']].min())
print(df[['na_sales', 'eu_sales', 'jp_sales', 'other_sales', 'critic_score', 'user_score']].max())

na_sales         0.0
eu_sales         0.0
jp_sales         0.0
other_sales      0.0
critic_score    13.0
dtype: object
na_sales          41.36
eu_sales        unknown
jp_sales        unknown
other_sales       10.57
critic_score       98.0
dtype: object


In [19]:
print(df['user_score'])

0          8
1        NaN
2        8.3
3          8
4        NaN
        ... 
16951    NaN
16952    NaN
16953    NaN
16954    NaN
16955    NaN
Name: user_score, Length: 16954, dtype: object


In [20]:
#df['user_score'] = pd.to_numeric(df['user_score'], downcast='float', errors='raise')
#print(df['user_score'])

ну видно что нужно оценить количестов выыбросов во всех строчках где type object и что-то с этим сделать аообще в задании написано что можно заменить все на пропуски, попробуем

In [21]:
df_new = df.copy()
df_new['eu_sales'] = pd.to_numeric(df_new['eu_sales'], downcast='float', errors='coerce')
df_new['jp_sales'] = pd.to_numeric(df_new['jp_sales'], downcast='float', errors='coerce')
df_new['user_score'] = pd.to_numeric(df_new['user_score'], downcast='float', errors='coerce')
print(df_new[['na_sales', 'eu_sales', 'jp_sales', 'other_sales', 'critic_score', 'user_score']].min())
print(df_new[['na_sales', 'eu_sales', 'jp_sales', 'other_sales', 'critic_score', 'user_score']].max())
df_new.info()
df.info()

na_sales         0.0
eu_sales         0.0
jp_sales         0.0
other_sales      0.0
critic_score    13.0
user_score       0.0
dtype: float64
na_sales        41.360000
eu_sales        28.959999
jp_sales        10.220000
other_sales     10.570000
critic_score    98.000000
user_score       9.700000
dtype: float64
<class 'pandas.core.frame.DataFrame'>
Int64Index: 16954 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16954 non-null  object 
 2   year_of_release  16954 non-null  int16  
 3   genre            16954 non-null  object 
 4   na_sales         16954 non-null  float64
 5   eu_sales         16948 non-null  float32
 6   jp_sales         16950 non-null  float32
 7   other_sales      16954 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       7688 non-null   float32
 10  rating           10085

In [22]:
print('eu_sales_old')
print(round(df['eu_sales'].isna().sum()/ df.shape[0] * 100, 2))
print('eu_sales_new')
print(round(df_new['eu_sales'].isna().sum()/ df.shape[0] * 100, 2))
print('jp_sales_old')
print(round(df['jp_sales'].isna().sum()/ df.shape[0] * 100, 2))
print('jp_sales_new')
print(round(df_new['jp_sales'].isna().sum()/ df.shape[0] * 100, 2))
print('user_score_old')
print(round(df['user_score'].isna().sum()/ df.shape[0] * 100, 2))
print('user_score_new')
print(round(df_new['user_score'].isna().sum()/ df.shape[0] * 100, 2))

eu_sales_old
0.0
eu_sales_new
0.04
jp_sales_old
0.0
jp_sales_new
0.02
user_score_old
40.12
user_score_new
54.65


можно заметить что вот в столбцах eu_sales и jp_sales содержание выбросов меньше 0.1 процента плюс ко всему не то чтобы сильно эти показатели были бы нужны для анализа в рамках нашего задания, удаляем, или даже вообще проигнорируем пусть будут посмотрим послее фильтррации однако оценки пользователей сильно уж много выбросов хочется узнать причину

In [23]:
df['eu_sales'] = pd.to_numeric(df['eu_sales'], downcast='float', errors='coerce')
df['jp_sales'] = pd.to_numeric(df['jp_sales'], downcast='float', errors='coerce')

In [24]:
print('разница после заполния всего неепонятного пропусками при приведении к типу')
print(df['user_score'].isna().sum() - df_new['user_score'].isna().sum())
new_ser = df[~df['user_score'].isna()]
print('столбики с tdb')
new_ser[new_ser['user_score'] == 'tbd'].info()


разница после заполния всего неепонятного пропусками при приведении к типу
-2464
столбики с tdb
<class 'pandas.core.frame.DataFrame'>
Int64Index: 2464 entries, 119 to 16947
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             2464 non-null   object 
 1   platform         2464 non-null   object 
 2   year_of_release  2464 non-null   int16  
 3   genre            2464 non-null   object 
 4   na_sales         2464 non-null   float64
 5   eu_sales         2464 non-null   float32
 6   jp_sales         2463 non-null   float32
 7   other_sales      2464 non-null   float64
 8   critic_score     1099 non-null   float64
 9   user_score       2464 non-null   object 
 10  rating           2411 non-null   object 
dtypes: float32(2), float64(3), int16(1), object(5)
memory usage: 197.3+ KB


как видим 2464 строчки имеют неопределенный рейтинг tbd ну считай пропуск смело заменяем на пропуск по факут получяем

In [25]:
df['user_score'] = pd.to_numeric(df['user_score'], downcast='float', errors='coerce')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16954 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16954 non-null  object 
 2   year_of_release  16954 non-null  int16  
 3   genre            16954 non-null  object 
 4   na_sales         16954 non-null  float64
 5   eu_sales         16948 non-null  float32
 6   jp_sales         16950 non-null  float32
 7   other_sales      16954 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       7688 non-null   float32
 10  rating           10085 non-null  object 
dtypes: float32(3), float64(3), int16(1), object(4)
memory usage: 1.3+ MB


вот получается что привели все столбики паралельно к нужному типу

### 2.4. Явные и неявные дубликаты в данных

In [26]:
platforms = df['platform'].unique()
platforms_count = df['platform'].nunique()
print(platforms)
print(platforms_count)

['Wii' 'NES' 'GB' 'DS' 'X360' 'PS3' 'PS2' 'SNES' 'GBA' 'PS4' '3DS' 'N64'
 'PS' 'XB' 'PC' '2600' 'PSP' 'XOne' 'WiiU' 'GC' 'GEN' 'DC' 'PSV' 'SAT'
 'SCD' 'WS' 'NG' 'TG16' '3DO' 'GG' 'PCFX']
31


Для палтформ неявных дубликатов не обнаружил всего 31 уникальне значение

In [27]:
genres = df['genre'].unique()
genres_count = df['genre'].nunique()
print(genres)
print(genres_count)

['Sports' 'Platform' 'Racing' 'Role-Playing' 'Puzzle' 'Misc' 'Shooter'
 'Simulation' 'Action' 'Fighting' 'Adventure' 'Strategy' 'MISC'
 'ROLE-PLAYING' 'RACING' 'ACTION' 'SHOOTER' 'FIGHTING' 'SPORTS' 'PLATFORM'
 'ADVENTURE' 'SIMULATION' 'PUZZLE' 'STRATEGY']
24


ага, неявные дубликатыы в жанрах обнаружены, переведем все в нижний регистр

In [28]:
df['genre'] = df['genre'].str.lower()
genres = df['genre'].unique()
genres_count = df['genre'].nunique()
print(genres)
print(genres_count)

['sports' 'platform' 'racing' 'role-playing' 'puzzle' 'misc' 'shooter'
 'simulation' 'action' 'fighting' 'adventure' 'strategy']
12


неявные дубликаты устранены для жанров устранены

In [29]:
ratings = df['rating'].unique()
ratings_count = df['rating'].nunique()
print(ratings)
print(ratings_count)

['E' nan 'M' 'T' 'E10+' 'K-A' 'AO' 'EC' 'RP']
8


Спрятался неявный дубликат 'K-A' то же самое что и 'E'

In [30]:
df['rating'] = df['rating'].str.replace('K-A', 'E')
ratings = df['rating'].unique()
ratings_count = df['rating'].nunique()
print(ratings)
print(ratings_count)

['E' nan 'M' 'T' 'E10+' 'AO' 'EC' 'RP']
7


Теперь и здесь нет неявных дубликатов

In [31]:
ratings = df['year_of_release'].unique()
ratings_count = df['year_of_release'].nunique()
print(sorted(ratings))
print(ratings_count)

[-1, 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016]
38


Все нормально, ну было бы странно но и чего то необычного тоже нет

С имменами все сложнее просто пока переведем их к нижнему регистру

In [32]:
print(df['name'].nunique())
print(df.shape[0])

11559
16954


In [33]:
df['name'] = df['name'].str.lower()

In [34]:
print(df['name'].nunique())
print(df.shape[0])
print(sorted(df['name'].unique()))

11559
16954
[' beyblade burst', ' fire emblem fates', " frozen: olaf's quest", ' haikyu!! cross team match!', ' tales of xillia 2', "'98 koshien", '.hack//g.u. vol.1//rebirth', '.hack//g.u. vol.2//reminisce', '.hack//g.u. vol.2//reminisce (jp sales)', '.hack//g.u. vol.3//redemption', '.hack//infection part 1', '.hack//link', '.hack//mutation part 2', '.hack//outbreak part 3', '.hack//quarantine part 4: the final chapter', '.hack: sekai no mukou ni + versus', '007 racing', '007: quantum of solace', '007: the world is not enough', '007: tomorrow never dies', '1 vs. 100', '1/2 summer +', '10 minute solution', '100 all-time favorites', '100 classic books', '100 classic games', '1000 cooking recipes from elle à table', '1001 touch games', '101-in-1 explosive megamix', '101-in-1 party megamix wii', '101-in-1 sports megamix', '101-in-1 sports party megamix', '1080°: teneighty snowboarding', '11eyes: crossover', '12-sai. honto no kimochi', '12-sai. koisuru diary', '12riven: the psi-climinal of

Можно заметить что вне зависимости от регистра очень много повторов на 11559 уникальныых названий приходится 16954 всего строк удалим сначало сто процентно одинаковые строки

In [35]:
df = df.drop_duplicates()
print(df['name'].nunique())
print(df.shape[0])

11559
16713


Стало немнного меньше

In [36]:
print(df.duplicated(subset =['name' , 'genre', 'platform','year_of_release' ]).sum())
print(df[df.duplicated(subset =['name' , 'genre', 'platform','year_of_release' ], keep=False)])
print(df.shape[0])


1
                name platform  year_of_release   genre  na_sales  eu_sales  \
606    madden nfl 13      PS3             2012  sports      2.11      0.22   
16465  madden nfl 13      PS3             2012  sports      0.00      0.01   

       jp_sales  other_sales  critic_score  user_score rating  
606         0.0         0.23          83.0         5.5      E  
16465       0.0         0.00          83.0         5.5      E  
16713


С учтом того что всякие количесвтеныые метрики отностительно грязные на предмет пропусков, найдем дубликаты по всем строчкам категорий название жанр, платформа, и год релиза все то что должно точно характеризовать видео игру видим тчо такой повтор только один отавим первый дубликат а последний удалим

In [37]:
df.drop_duplicates(subset =['name' , 'genre', 'platform','year_of_release' ], keep='first',inplace=True)
print(df.duplicated(subset =['name' , 'genre', 'platform','year_of_release' ]).sum())
print(df[df.duplicated(subset =['name' , 'genre', 'platform','year_of_release' ], keep=False)])
print(df.shape[0])

0
Empty DataFrame
Columns: [name, platform, year_of_release, genre, na_sales, eu_sales, jp_sales, other_sales, critic_score, user_score, rating]
Index: []
16712


У нас есть пробелы в дате релиза может сможем заполнить их повторящимися элементами

In [38]:
print(df.duplicated(subset =['name' , 'genre', 'platform']).sum())
print(df[df.duplicated(subset =['name' , 'genre', 'platform'], keep=False)])
print(df.shape[0])

3
                              name platform  year_of_release     genre  \
1197   need for speed: most wanted     X360             2012    racing   
1605   need for speed: most wanted     X360             2005    racing   
1760            sonic the hedgehog      PS3             2006  platform   
4173            sonic the hedgehog      PS3               -1  platform   
6051   need for speed: most wanted       PC             2005    racing   
11873  need for speed: most wanted       PC             2012    racing   

       na_sales  eu_sales  jp_sales  other_sales  critic_score  user_score  \
1197       0.62      0.78      0.01         0.15          83.0         8.5   
1605       1.00      0.13      0.02         0.10          83.0         8.5   
1760       0.41      0.06      0.04         0.66          43.0         4.1   
4173       0.00      0.48      0.00         0.00          43.0         4.1   
6051       0.02      0.23      0.00         0.04          82.0         8.5   
11873      

sonic the hedgehog - одна гипотеза себя оправдала, удалим ту что без года. что касается need for speed: most wantedneed for speed: most wanted честно изучив вопрос поглубже узнал что у игры need for speed: most wanted был ремастер в 2012 году и похорошему надо бы изменить имя у той что вышла в 2012 для каждой из платформ на ремастер, но может это просто ошибка, рейтинг одинаковый поменям просто в ручную

In [39]:
df.loc[[1197, 11873], 'name'] = 'need for speed: most wanted remaster'
print(df.loc[[1197,1605,6051, 11873], 'name'])
df.drop_duplicates(subset =['name' , 'genre', 'platform'], keep='first',inplace=True)
print(df.duplicated(subset =['name' , 'genre', 'platform']).sum())
print(df[df.duplicated(subset =['name' , 'genre', 'platform'], keep=False)])
print(df.shape[0])

1197     need for speed: most wanted remaster
1605              need for speed: most wanted
6051              need for speed: most wanted
11873    need for speed: most wanted remaster
Name: name, dtype: object
0
Empty DataFrame
Columns: [name, platform, year_of_release, genre, na_sales, eu_sales, jp_sales, other_sales, critic_score, user_score, rating]
Index: []
16711


Рассуждаем дальше как бы норамльно что есть игры которые были выыпущеныы на разныых площадках, но при этом рейтинг должен быть без дубликатов, проверим

In [40]:
print(df.duplicated(subset =['name' , 'platform']).sum())
print(df[df.duplicated(subset =['name' , 'platform'], keep=False)])
print(df.shape[0])

0
Empty DataFrame
Columns: [name, platform, year_of_release, genre, na_sales, eu_sales, jp_sales, other_sales, critic_score, user_score, rating]
Index: []
16711


что и требовалось доказать теперь взгляенм под другим углом

In [41]:
print(df.duplicated(subset =['name' , 'genre']).sum())
print(df[df.duplicated(subset =['name' , 'genre'], keep=False)])
print(df.shape[0])

5146
                                                   name platform  \
1                                     super mario bros.      NES   
5                                                tetris       GB   
16                                   grand theft auto v      PS3   
17                        grand theft auto: san andreas      PS2   
18                                    super mario world     SNES   
...                                                 ...      ...   
16947                     men in black ii: alien escape       GC   
16950  score international baja 1000: the official game      PS2   
16951                     samurai warriors: sanada maru      PS3   
16954                                  spirits & spells      GBA   
16955                               winning post 8 2016      PSV   

       year_of_release       genre  na_sales  eu_sales  jp_sales  other_sales  \
1                 1985    platform     29.08      3.58      6.81         0.77   
5               

Это нормально что для одной и той же игры есть разные площадки выхода поэтому это уже не дубликаты соответственно

ВЫВОДЫ

Было 16956 строк Стало 16711 строк Удалено 245 строк это 0,014 от первоначального количества менее двух процентов Приэтом не были удаленыстроки с пропущеннымм годом в теории можно удалить ибо меньше дввух процентов, но мможно индивидульно прогуглить каждую игру и ее год на этапе фильтрации мыы избавимся от строк с незааполненными датами

In [42]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16711 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16711 non-null  object 
 1   platform         16711 non-null  object 
 2   year_of_release  16711 non-null  int16  
 3   genre            16711 non-null  object 
 4   na_sales         16711 non-null  float64
 5   eu_sales         16705 non-null  float32
 6   jp_sales         16707 non-null  float32
 7   other_sales      16711 non-null  float64
 8   critic_score     8135 non-null   float64
 9   user_score       7588 non-null   float32
 10  rating           9947 non-null   object 
dtypes: float32(3), float64(3), int16(1), object(4)
memory usage: 1.2+ MB


О чудо а итоге без удалений мы избавилирсь от всех игр без даты релиза, так и задумывыалось?

In [43]:
print(df[df['eu_sales'].isna()])
print(df[df['jp_sales'].isna()])

                                  name platform  year_of_release     genre  \
446                      rhythm heaven       DS             2008      misc   
802                        dead rising     X360             2006    action   
1131  prince of persia: warrior within      PS2             2004    action   
1132                         far cry 4     XOne             2014   shooter   
1394                   sonic advance 3      GBA             2004  platform   
1612                       ratatouille       DS             2007    action   

      na_sales  eu_sales  jp_sales  other_sales  critic_score  user_score  \
446       0.55       NaN      1.93         0.13          83.0         9.0   
802       1.16       NaN      0.08         0.20          85.0         7.6   
1131      0.54       NaN      0.00         0.22          83.0         8.5   
1132      0.80       NaN      0.01         0.14          82.0         7.5   
1394      0.74       NaN      0.08         0.06          79.0       

Их осталось так мало, быть может стоит заполнить эти ячейки значением 0.0, если предполагать что nan отсуствие продаж вообще ну не пустили на рынок в виду ограничеений значений так мало плюс они явно не повлияют на дальнейшее исследование

In [44]:
df['jp_sales']=df['jp_sales'].fillna(0.0)
df['eu_sales']=df['eu_sales'].fillna(0.0)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16711 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16711 non-null  object 
 1   platform         16711 non-null  object 
 2   year_of_release  16711 non-null  int16  
 3   genre            16711 non-null  object 
 4   na_sales         16711 non-null  float64
 5   eu_sales         16711 non-null  float32
 6   jp_sales         16711 non-null  float32
 7   other_sales      16711 non-null  float64
 8   critic_score     8135 non-null   float64
 9   user_score       7588 non-null   float32
 10  rating           9947 non-null   object 
dtypes: float32(3), float64(3), int16(1), object(4)
memory usage: 1.2+ MB


К фильтрации!!!

---

## 3. Фильтрация данных

новый срез данных (период с 2000 по 2013 год включительно) в отдельном датафрейме

In [45]:
df_actual = df[(df['year_of_release'].between(2000, 2013))].copy()
df_actual.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12780 entries, 0 to 16954
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             12780 non-null  object 
 1   platform         12780 non-null  object 
 2   year_of_release  12780 non-null  int16  
 3   genre            12780 non-null  object 
 4   na_sales         12780 non-null  float64
 5   eu_sales         12780 non-null  float32
 6   jp_sales         12780 non-null  float32
 7   other_sales      12780 non-null  float64
 8   critic_score     7168 non-null   float64
 9   user_score       6482 non-null   float32
 10  rating           8722 non-null   object 
dtypes: float32(3), float64(3), int16(1), object(4)
memory usage: 973.5+ KB


In [46]:
df_actual['year_of_release'].unique()

array([2006, 2008, 2009, 2005, 2007, 2010, 2013, 2004, 2002, 2001, 2011,
       2012, 2003, 2000], dtype=int16)

---

## 4. Категоризация данных
    
Произведена категоризация данных:
- Разделены все игры по оценкам пользователей и выдеены такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [47]:
df_actual.sort_values(by=['name', 'platform']).head(50)

name platform  \
3394                                frozen: olaf's quest      3DS   
3906                                frozen: olaf's quest       DS   
2478                                   tales of xillia 2      PS3   
8460                          .hack//g.u. vol.1//rebirth      PS2   
7182                        .hack//g.u. vol.2//reminisce      PS2   
8719             .hack//g.u. vol.2//reminisce (jp sales)      PS2   
8410                       .hack//g.u. vol.3//redemption      PS2   
1575                             .hack//infection part 1      PS2   
9189                                         .hack//link      PSP   
3023                              .hack//mutation part 2      PS2   
4313                              .hack//outbreak part 3      PS2   
8098         .hack//quarantine part 4: the final chapter      PS2   
14525                  .hack: sekai no mukou ni + versus      PS3   
3800                                          007 racing       PS   
9637                              007: quantum of solace       DS   
14860                             007: quantum of solace       PC   
4524                              007: quantum of solace      PS2   
1802                              007: quantum of solace      PS3   
3153                              007: quantum of solace      Wii   
1293                              007: quantum of solace     X360   
1211                        007: the world is not enough      N64   
2263                        007: the world is not enough       PS   
11425                                          1 vs. 100       DS   
16455                                       1/2 summer +      PSP   
11878                                 10 minute solution      Wii   
3942                              100 all-time favorites       DS   
3087                                   100 classic books       DS   
14328                                  100 classic games       DS   
15010             1000 cooking recipes from elle à table       DS   
5799                                    1001 touch games       DS   
7887                          101-in-1 explosive megamix       DS   
7236                          101-in-1 party megamix wii      Wii   
11541                            101-in-1 sports megamix       DS   
14982                      101-in-1 sports party megamix      Wii   
15880                                  11eyes: crossover      PSP   
15301                                  11eyes: crossover     X360   
13924              12riven: the psi-climinal of integral      PS2   
16946                                            15 days       PC   
6003                                           1701 a.d.       PC   
10216                   18 wheeler: american pro trucker       GC   
4860                    18 wheeler: american pro trucker      PS2   
15504                18 wheels of steel: extreme trucker       PC   
10143              18 wheels of steel: extreme trucker 2       PC   
9003                                    187: ride or die      PS2   
12450                                   187: ride or die       XB   
9971   2 games in 1 double pack: hot wheels velocity ...      GBA   
1008   2 games in 1 double pack: the incredibles / fi...      GBA   
1024       2 games in 1: disney princess & the lion king      GBA   
5824   2 games in 1: disney's brother bear / the lion...      GBA   
955         2 games in 1: sonic advance & chuchu rocket!      GBA   

       year_of_release         genre  na_sales  eu_sales  jp_sales  \
3394              2013      platform      0.27      0.27      0.00   
3906              2013      platform      0.21      0.26      0.00   
2478              2012  role-playing      0.20      0.12      0.45   
8460              2006  role-playing      0.00      0.00      0.17   
7182              2006  role-playing      0.11      0.09      0.00   
8719              2006  role-playing      0.00      0.00      0.16   
8410              2007  role-playing      

Ну блин, что делать с пропусками в рейтинге не понятно их слишком много, категоризировать игру на определенной платформе по рейтингу на другой не до конца правильно, искать какое то среднее на других площадках для этой игры тоже неверно, это с условием того что игра может не иметь рейтинга ни на одной из площадок, или быть выпущена тольокк на одной платформе и не иметь рейтинга.

При этом ну неверно убирать так много данных

Было принято решение выделить четвертую категориюне имеет оценки для того чтобы оставить как можно большее строк в датасете ведь игры были выпущениы и удаление строк повлияет на результат при получении топ-7 платформ, при этом оставить разделение на категории максимально объективным

In [48]:
def user_rate_to_cat(row):
    if row['user_score'] <= 10 and row['user_score'] >= 8:
        return "высокая оценка"
    elif  row['user_score'] < 8 and row['user_score'] >= 3:
        return "средняя оценка"
    elif  row['user_score'] < 3 and row['user_score'] >= 0:
        return "низкая оценка"
    else: return "нет оценки"
df_actual['category_user'] = df_actual.apply(user_rate_to_cat, axis=1)
df_actual.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12780 entries, 0 to 16954
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             12780 non-null  object 
 1   platform         12780 non-null  object 
 2   year_of_release  12780 non-null  int16  
 3   genre            12780 non-null  object 
 4   na_sales         12780 non-null  float64
 5   eu_sales         12780 non-null  float32
 6   jp_sales         12780 non-null  float32
 7   other_sales      12780 non-null  float64
 8   critic_score     7168 non-null   float64
 9   user_score       6482 non-null   float32
 10  rating           8722 non-null   object 
 11  category_user    12780 non-null  object 
dtypes: float32(3), float64(3), int16(1), object(5)
memory usage: 1.0+ MB


In [49]:
def critic_rate_to_cat(row):
    if row['critic_score'] <= 100 and row['critic_score'] >= 80:
        return "высокая оценка"
    elif  row['critic_score'] < 80 and row['critic_score'] >= 30:
        return "средняя оценка"
    elif  row['critic_score'] < 30 and row['critic_score'] >= 0:
        return "низкая оценка"
    else: return "нет оценки"
df_actual['category_critic'] = df_actual.apply(critic_rate_to_cat, axis=1)
df_actual.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12780 entries, 0 to 16954
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             12780 non-null  object 
 1   platform         12780 non-null  object 
 2   year_of_release  12780 non-null  int16  
 3   genre            12780 non-null  object 
 4   na_sales         12780 non-null  float64
 5   eu_sales         12780 non-null  float32
 6   jp_sales         12780 non-null  float32
 7   other_sales      12780 non-null  float64
 8   critic_score     7168 non-null   float64
 9   user_score       6482 non-null   float32
 10  rating           8722 non-null   object 
 11  category_user    12780 non-null  object 
 12  category_critic  12780 non-null  object 
dtypes: float32(3), float64(3), int16(1), object(6)
memory usage: 1.1+ MB


Категории по ррейтингу пользователей

In [50]:
group_user = df_actual.groupby('category_user')['name'].agg('count').sort_values(ascending=False)
print(group_user) 

category_user
нет оценки        6298
средняя оценка    4080
высокая оценка    2286
низкая оценка      116
Name: name, dtype: int64


Категории по рейтингу критиков

In [51]:
group_critic = df_actual.groupby('category_critic')['name'].agg('count').sort_values(ascending=False)
print(group_critic) 

category_critic
нет оценки        5612
средняя оценка    5422
высокая оценка    1691
низкая оценка       55
Name: name, dtype: int64


топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [52]:
top_7 = df.groupby('platform')['name'].agg('count').sort_values(ascending=False).head(7)
print(top_7) 

platform
PS2     2161
DS      2151
PS3     1329
Wii     1320
X360    1262
PSP     1209
PS      1197
Name: name, dtype: int64


---

## 5. Итоговый вывод



В первую очередь быыла проделана огромная работа по предобработке данных К общему стилю snake_case были приведены названия столбцов Далее посккольку тип серии данных напрямую зависит от пропусков данных паралельно совершалось заполнение пропусков, а также приведение к нужному для работы типу Некоторые строки были удалены за неимением смысла, что-то заменили индикатором (но в итоге удалилось все само), что-то оставили без изменения (что привело к появлению новой категории без рейтинга на этапе категоризации Был проведен глубокий анализ на наличие дубликатов как явных такк и не явных, Было 16956 строк Стало 16711 строк Удалено 245 строк это 0,014 от первоначального количества менее двух процентов Далее мы отфильтровали данные в срезе нужных нам годов с 2000 по 2013 Провели категоризацию не по 3 а по четыырем категориям высокая оценка, средня, низка, без оценки в результате чего появилось два новых поля отражающих категории для двух оценок Последжним этапом было составление топ-7 платформ по выпуску игр на данном временном интревале, перввое место достается PS2 для которой было выпущено 2127 игр
Также очень много не оцененных игр, большая часть в основном игры имеют среедний рейтинг, очень мало выско оцененных игр, еще меньше нипзкооцененных